### DISCUSSION TITLE
"connection refused" on mox for R Studio server

### DISCUSSION THREAD

**Discussion by yaaminiv** [RobertsLab/resources#1190] [Q&A] [answered]
I'm following @kubu4's instructions to access the R Studio server on mox [in this discussion](https://github.com/RobertsLab/resources/discussions/1180#discussioncomment-651295). I am trying to tunnel into mox with the following command:

> ssh -N -L 8787:n2221:33264 yaaminiv@mox.hyak.uw.edu

I'm able to log into mox, but I get the following "connection refused" error:

<img width="605" alt="Screen Shot 2021-04-27 at 1 02 18 PM" src="https://user-images.githubusercontent.com/22335838/116305568-cf313000-a758-11eb-81b8-47af5c376f10.png">

Subsequently I cannot access http://localhost:8787/. Is there a reason why the connection is refused?

**Comment by kubu4** [comment, marked as answer]
When I check the Mox queue, your script isn't currently running. That'd be my guess.

**Reply by yaaminiv** [reply]
Here's my script: https://github.com/RobertsLab/project-gigas-oa-meth/blob/master/code/06-methylKit-RStudio.sh

I copied it from yours but didn't include the part at the bottom to set a temporary directory, since I wanted to put everything in the directory specified by `--chdir`. Is there something I need to specify so the script doesn't end?

**Reply by kubu4** [reply]
Your script is missing all of this:

```
mkdir -p "$TMPDIR/tmp/rstudio-server"
uuidgen > "$TMPDIR/tmp/rstudio-server/secure-cookie-key"
chmod 0600 "$TMPDIR/tmp/rstudio-server/secure-cookie-key"

mkdir -p "$TMPDIR/var/lib"
mkdir -p "$TMPDIR/var/run"

# User-installed R packages go into their home directory
if [ ! -e ${HOME}/.Renviron ]
then
  printf '\nNOTE: creating ~/.Renviron file\n\n'
  echo 'R_LIBS_USER=~/R/%p-library/%v' >> ${HOME}/.Renviron
fi

# This example bind mounts the /project directory on the host into the Singularity container.
# By default the only host file systems mounted within the container are $HOME, /tmp, /proc, /sys, and /dev.
singularity exec \
--bind="$TMPDIR/var/lib:/var/lib/rstudio-server" \
--bind="$TMPDIR/var/run:/var/run/rstudio-server" \
--bind="$TMPDIR/tmp:/tmp" \
--bind=/gscratch/scrubbed/${USER} \
/gscratch/srlab/programs/singularity_containers/rstudio-4.0.2.sjw-01 \
rserver --www-port ${PORT} --auth-none=0 --auth-pam-helper-path=pam-helper
```

_All_ of that is required. Particularly because a big chunk of that is what actually starts RStudio Server. Also, don't modify any of it. :)

**Reply by yaaminiv** [reply]
OHHHHH when I quickly read your script I thought all of that was to set a temporary directory and didn't look at it again! I added that chunk above to [my script](https://github.com/RobertsLab/project-gigas-oa-meth/blob/master/code/06-methylKit-RStudio.sh) and am now able to log into R Studio server!

![Screen Shot 2021-04-27 at 1 26 29 PM](https://user-images.githubusercontent.com/22335838/116308307-2e447400-a75c-11eb-9138-99da18be4f18.png)

If I close my computer, will the R Studio session end?

**Reply by kubu4** [reply]
Great

I believe that as long as the SLURM script is running, you can come and go as you please; the current session should remain.

**Reply by yaaminiv** [reply]
> I believe that as long as the SLURM script is running, you can come and go as you please; the current session should remain.

Can confirm that this is NOT the case

<img width="733" alt="Screen Shot 2021-04-27 at 10 34 46 PM" src="https://user-images.githubusercontent.com/22335838/116351578-c6684a80-a7a8-11eb-8655-b981b8db00dd.png">

**Reply by kubu4** [reply]
Ooof, this is no good... I'll look into this and see if I can come up with any ideas.

**Reply by kubu4** [reply]
Can you describe what you did to get these error messages? I just did some testing (where I ran the `Sys.sleep()`, followed by a print statement).

1. Closed the browser.

2. Killed the SSH tunnel.

3. After the set time in the `Sys.sleep()` command (120 seconds), I re-established the SSH connection and re-opened the browser. 

The job completed without errors.

**Reply by kubu4** [reply]
Doing some poking around, is it possible your home directory quota is used up? This is my suspicion. It looks like RStudio Server is set up to save the Global Environment info in `~/.rstudio`. If you're methylkit commands are generating large objects (which I would expect), I'm guessing that these are trying to get saved in `~/.rstudio` which is using up all of your quota in your home directory.

The above is based on this article:

https://support.rstudio.com/hc/en-us/articles/218417097-Filling-up-the-home-directory-with-RStudio-Server

I'll see if I can update the container to get around this issue.

**Reply by kubu4** [reply]
Also, if space usage on your login node is the issue, it could be related to your `~/.Rdata` file. Did you elect to save this file (if/when prompted by RStudio)? If yes, this, too, could end up being a very large file, depending on what you're doing/running.

**Reply by kubu4** [reply]
_Also_, as it turns out, the `srlab` is over our Mox quota! This could also be causing issues... Will try to free up some space.

**Reply by yaaminiv** [reply]
> Can you describe what you did to get these error messages? I just did some testing (where I ran the Sys.sleep(), followed by a print statement).

All I did was close my browser window with R Studio server! That ended up killing my SSH tunnel (totally forgot to take a screenshot). When I logged in again, I saw that error message in R Studio. Maybe I did not  follow the correct protocol for closing a session to still have it run?

**Comment by yaaminiv** [comment]
> Doing some poking around, is it possible your home directory quota is used up? This is my suspicion.

I don't know if my home directory quota is used up! How do I check?

Also, I set my working directory to be `/gscratch/scrubbed/yaaminiv/Manchester/analyses/methylKit` when I started my R Studio session. Do I need to set up R Studio server to save `~/.rstudio` in that folder instead? 

> Also, if space usage on your login node is the issue, it could be related to your `~/.Rdata` file. Did you elect to save this file (if/when prompted by RStudio)? 

I have not elected to save my `~/.Rdata`, and manually write out my data to `/gscratch/scrubbed/yaaminiv/Manchester/analyses/methylKit/methylKit.Rdata` instead!

**Comment by yaaminiv** [comment]
Last night I set up R Studio server to run on genefish since closing my browser window killed my R Studio session. I just logged onto genefish to check my R Studio server session. I found that I had been automatically logged out due to 60 min of inactivity! When I logged back in, I saw the several error messages:

![Screen Shot 2021-04-28 at 11 26 07 AM](https://user-images.githubusercontent.com/22335838/116454261-a4a4ad00-a814-11eb-8ccb-c89cf5516c4a.png)

1. `There were 50 or more warnings (use warnings() to see the first 50)`: Checked warnings(), they're glm.fit errors (not related to the issue at hand)

![Screen Shot 2021-04-28 at 11 26 30 AM](https://user-images.githubusercontent.com/22335838/116454391-caca4d00-a814-11eb-989c-26071f27073d.png)

2. `Error in h(simpleError(msg, call)) : 
  error in evaluating the argument 'x' in selecting a method for function 'head': object 'differentialMethylationStatsTreatment' not found`: My fault, didn't reference the correct data frame

3.  Same `error writing to connection` I posted above! However, it seems like my `calculateDiffMeth` command finished running, and my global environment was loaded again when I logged back into the R Studio server? So now I'm just not sure what these errors mean...

**Reply by kubu4** [reply]
Check quota:

`mmlsquota --block-size G gscratch:home`

Also, from a build node, please run the following commands and copy/paste results:

```shell
cd
du -hs .[^.]*
````

**Reply by yaaminiv** [reply]
> Check quota

![Screen Shot 2021-04-28 at 1 50 35 PM](https://user-images.githubusercontent.com/22335838/116470777-b5f7b480-a828-11eb-83ad-063ab7574535.png)

**Comment by yaaminiv** [comment]
> Also, from a build node, please run the following commands and copy/paste results

128K	.bash_history
0	.bash_profile
0	.bashrc
1.7M	.cache
0	.conda
0	.condarc
0	.config
0	.cshrc
0	.gitconfig
640K	.java
3.9M	.local
0	.login
0	.pki
0	.Renviron
0	.rnd
2.0M	.rstudio
0	.singularity
0	.ssh
0	.Xauthority

**Reply by kubu4** [reply]
Cool, ruled out storage space in `${HOME}` as a potential culprit. Thanks!

We're still dealing with exceeding quota on `/gscratch/srlab` ATM. This could be the issue. Will report back once that's resolved...

**Reply by kubu4** [reply]
Mox storage has been expanded and we're no longer over quota. Give it another rip when you have a chance.
